### PR TITLE
Show diff against the messages from the firebase

### PR DESCRIPTION
**Pull request by elayabharath**

### CODE REVIEW

File: webapp/lib/app/configurator/standard_messages/controller.dart [hunk 1]
```diff
@@ -195,34 +188,21 @@ class MessagesConfiguratorController extends ConfiguratorController {
...
-      }
-      otherPartSaved = true;
-    }, onError: (error, stacktrace) {
-      _view.showSaveStatus('Unable to save. Please check your connection and try again. If the issue persists, please contact your project administrator');
-    });
+    _view.disableSaveButton();
+    var diffData = standardMessagesManager.diffData;
+
+    try {
+      await Future.wait([platform.updateSuggestedReplies(diffData.editedMessages), platform.deleteSuggestedReplies(diffData.deletedMessages)]);
```
Comment (marianamarasoiu) [inline webapp/lib/app/configurator/standard_messages/controller.dart:197]: nice, thanks for the refactor!

File: webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart [hunk 2]
```diff
@@ -24,9 +37,79 @@ class StandardMessagesManager {
...
+    for (var category in fbCategories.values) {
+      category.messages.forEach((message) {
+        fbMessages[message.suggestedReplyId] = message;
+      });
+    }
+
+    Set<String> allMessageKeys = Set()..addAll(messages.keys)..addAll(fbMessages.keys);
+    for (var messageId in allMessageKeys) {
+      var fbMessage = fbMessages[messageId];
+      var localMessage = messages[messageId];
```
Comment (marianamarasoiu) [inline webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart:69]: I quite like `local` as a way of referring to the changes shown in the UI - at the risk of becoming a bit verbose, how about we use that everywhere? (e.g. `localCategories`, `localMessages` etc?)

`fb` is too reminiscent of facebook as a shortcut for firebase, so in the same vein as `local`, we could use the prefix `storage`?

File: webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart [hunk 3]
```diff
@@ -52,6 +135,23 @@ class StandardMessagesManager {
     return added;
   }
 
+  model.SuggestedReply onAddStandardMessageFromFb(model.SuggestedReply standardMessage) {
```
Comment (marianamarasoiu) [inline webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart:138]: I think the local/storage naming above can also be used here, so we get  addStorageMessage(s), and can work for update and remove as well.

the methods modifying the local data can use `local` instead, and so we can see the difference between them much easier too

File: webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart (hunk 2 above)
Comment (elayabharath) [reply webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart]: Makes sense, updated all the methods to local/storage where appropriate

File: webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart [hunk 4]
```diff
@@ -9,81 +22,184 @@ class StandardMessagesManager {
...
 
   int getNextCategoryIndex() {
-    var lastIndex = categories.values.fold(0, (previousValue, category) => previousValue > category.categoryIndex ? previousValue : category.categoryIndex);
+    var lastIndex = localCategories.values.fold(0, (previousValue, category) => previousValue > category.categoryIndex ? previousValue : category.categoryIndex);
     return lastIndex + 1;
   }
 
   int getNextGroupIndexInCategory(String categoryId) {
-    var lastIndex = categories[categoryId].groups.values.fold(0, (previousValue, group) => previousValue > group.groupIndexInCategory ? previousValue : group.groupIndexInCategory);
+    var lastIndex = localCategories[categoryId].groups.values.fold(0, (previousValue, group) => previousValue > group.groupIndexInCategory ? previousValue : group.groupIndexInCategory);
```
Comment (marianamarasoiu) [inline webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart:30]: nit:
```suggestion
    var lastIndex = localCategories[categoryId].groups.values.fold(0, (maxIndex, group) => maxIndex > group.groupIndexInCategory ? maxIndex : group.groupIndexInCategory);
```

File: webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart [hunk 5]
```diff
@@ -9,81 +22,184 @@ class StandardMessagesManager {
...
+    List<model.SuggestedReply> editedMessages = [];
+    List<model.SuggestedReply> deletedMessages = [];
+
+    Map<String, model.SuggestedReply> storageMessages = {};
+    Map<String, model.SuggestedReply> localMessages = {};
+
+    for (var category in localCategories.values) {
+      category.messages.forEach((message) {
+        localMessages[message.suggestedReplyId] = message;
+      });
```
Comment (marianamarasoiu) [inline webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart:57]: nit:
```suggestion
      for (var message in category.messages) {
        localMessages[message.suggestedReplyId] = message;
      }
```

File: webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart [hunk 6]
```diff
@@ -9,81 +22,184 @@ class StandardMessagesManager {
...
+    }
+
+    for (var category in storageCategories.values) {
+      category.messages.forEach((message) {
+        storageMessages[message.suggestedReplyId] = message;
+      });
```
Comment (marianamarasoiu) [inline webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart:63]: nit:
```suggestion
      for (var message in category.messages) {
        storageMessages[message.suggestedReplyId] = message;
      }
```

File: webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart [hunk 7]
```diff
@@ -9,81 +22,184 @@ class StandardMessagesManager {
...
+    }
+
+    Set<String> allMessageKeys = Set()..addAll(localMessages.keys)..addAll(storageMessages.keys);
```
Comment (marianamarasoiu) [inline webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart:66]: nit, there's a nice shortcut for making a list or set out of multiple lists/sets by prefixing them with `...` (you can also add individual items at the same time if needed, but it's not the case here)
```suggestion
    Set<String> allMessageKeys = {...localMessages.keys, ...storageMessages.keys};
```

File: webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart [hunk 8]
```diff
@@ -9,81 +22,184 @@ class StandardMessagesManager {
...
+  List<model.SuggestedReply> get standardMessagesInLocal => localCategories.values.fold([], (result, category) => result..addAll(category.messages));
 
-  model.SuggestedReply getStandardMessageById(String id) => standardMessages.singleWhere((r) => r.suggestedReplyId == id);
+  model.SuggestedReply getStandardMessageById(String id) => standardMessagesInLocal.singleWhere((r) => r.suggestedReplyId == id);
 
 
-  model.SuggestedReply addStandardMessage(model.SuggestedReply standardMessage) {
-    if (standardMessages.any((element) => element.suggestedReplyId == standardMessage.suggestedReplyId)) {
-      updateStandardMessage(standardMessage);
+  model.SuggestedReply addStandardMessageInLocal(model.SuggestedReply standardMessage) {
```
Comment (marianamarasoiu) [inline webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart:118]: nit - let's make this a private method: `_addStandardMessageInLocal`

File: webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart [hunk 9]
```diff
@@ -9,81 +22,184 @@ class StandardMessagesManager {
...
-    categories[standardMessage.categoryId].groups.putIfAbsent(standardMessage.groupId, () => MessageGroup(standardMessage.groupId, standardMessage.groupName, standardMessage.groupIndexInCategory));
-    categories[standardMessage.categoryId].groups[standardMessage.groupId].messages[standardMessage.suggestedReplyId] = standardMessage;
+    localCategories.putIfAbsent(standardMessage.categoryId, () => new MessageCategory(standardMessage.categoryId, standardMessage.category, standardMessage.categoryIndex));
+    localCategories[standardMessage.categoryId].groups.putIfAbsent(standardMessage.groupId, () => MessageGroup(standardMessage.groupId, standardMessage.groupName, standardMessage.groupIndexInCategory));
+    localCategories[standardMessage.categoryId].groups[standardMessage.groupId].messages[standardMessage.suggestedReplyId] = standardMessage;
     return standardMessage;
   }
 
-  List<model.SuggestedReply> addStandardMessages(List<model.SuggestedReply> standardMessagesToAdd) {
+  model.SuggestedReply onAddStandardMessageFromStorage(model.SuggestedReply standardMessage) {
```
Comment (marianamarasoiu) [inline webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart:129]: nit - let's rename this slightly and make it a private method: `_addStandardMessageInStorage`

File: webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart [hunk 10]
```diff
@@ -9,81 +22,184 @@ class StandardMessagesManager {
...
+      onAddStandardMessageFromStorage(standardMessage);
       if (result != null) added.add(result);
     }
     return added;
   }
 
-  model.SuggestedReply updateStandardMessage(model.SuggestedReply standardMessage) {
-    var removed = removeStandardMessage(standardMessage);
-    var updated = addStandardMessage(standardMessage);
+  model.SuggestedReply updateStandardMessageInLocal(model.SuggestedReply standardMessage) {
```
Comment (marianamarasoiu) [inline webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart:146]: nit - let's make this a private method: `_updateStandardMessageInLocal`

File: webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart [hunk 11]
```diff
@@ -9,81 +22,184 @@ class StandardMessagesManager {
...
     }
     return updated;
   }
 
-  List<model.SuggestedReply> updateStandardMessages(List<model.SuggestedReply> standardMessages) {
-    List<model.SuggestedReply> updated = [];
-    for (var standardMessage in standardMessages) {
-      var result = updateStandardMessage(standardMessage);
-      if (result != null) updated.add(result);
+  model.SuggestedReply onUpdateStandardMessageFromStorage(model.SuggestedReply standardMessage) {
```
Comment (marianamarasoiu) [inline webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart:155]: nit - let's rename this slightly and make it a private method: `_updateStandardMessageInStorage`

File: webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart [hunk 12]
```diff
@@ -9,81 +22,184 @@ class StandardMessagesManager {
...
+      var result = updateStandardMessageInLocal(standardMessage);
+      if (result != null) added.add(result);
     }
-    return updated;
+    return added;
   }
 
-  model.SuggestedReply removeStandardMessage(model.SuggestedReply standardMessage) {
-    if (!standardMessages.any((element) => element.suggestedReplyId == standardMessage.suggestedReplyId)) {
+  model.SuggestedReply removeStandardMessageInLocal(model.SuggestedReply standardMessage) {
```
Comment (marianamarasoiu) [inline webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart:175]: nit - let's make this a private method: `_removeStandardMessageInLocal`

File: webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart [hunk 13]
```diff
@@ -9,81 +22,184 @@ class StandardMessagesManager {
...
     }
-    var oldStandardMessage = standardMessages.singleWhere((element) => element.suggestedReplyId == standardMessage.suggestedReplyId);
-    categories[oldStandardMessage.categoryId].groups[oldStandardMessage.groupId].messages.remove(oldStandardMessage.suggestedReplyId);
+    var oldStandardMessage = standardMessagesInLocal.singleWhere((element) => element.suggestedReplyId == standardMessage.suggestedReplyId);
+    localCategories[oldStandardMessage.categoryId].groups[oldStandardMessage.groupId].messages.remove(oldStandardMessage.suggestedReplyId);
     return standardMessage;
   }
 
-  List<model.SuggestedReply> removeStandardMessages(List<model.SuggestedReply> standardMessages) {
+  model.SuggestedReply onRemoveStandardMessageFromStorage(model.SuggestedReply standardMessage) {
```
Comment (marianamarasoiu) [inline webapp/lib/app/configurator/standard_messages/controller_messages_helper.dart:186]: nit - let's rename this slightly and make it a private method: `_removeStandardMessageInStorage`

**Review by marianamarasoiu — APPROVED** [review]
thanks EB - LGTM with a few nits!

it feels like there are now separate methods for manipulating the local and storage data, so I've suggested that we rename these so that their names are similar. I think it would also be useful a quick follow up PR for moving these around so they're grouped together. Something like:

// methods for manipulating local data
_addStandardMessageInLocal
_updateStandardMessageInLocal
_removeStandardMessageInLocal

// methods for manipulating data from storage
_addStandardMessageInStorage
_updateStandardMessageInStorage
_removeStandardMessageInStorage

// methods to be called when updates are received from storage
onAddStandardMessagesFromStorage
onUpdateStandardMessagesFromStorage
onRemoveStandardMessagesFromStorage

// methods to be called when the user makes changes to the local configuration
createMessage
modifyMessage
deleteMessage

**Review by marianamarasoiu — APPROVED** [review]
LGTM, thanks for fixing the indicators @elayabharath !

**Pull request merged**